### PR TITLE
feat: Ensure experimental rest endpoint is present at startup

### PR DESCRIPTION
**Pull request by sadhansood**
This PR checks if the endpoints which returns full checkpoint is working at startup and fail with an appropriate error message.

### CODE REVIEW

File: crates/walrus-event/src/lib.rs [hunk 1]
```diff
@@ -196,3 +196,30 @@ pub async fn get_bootstrap_committee_and_checkpoint(
...
+        .get_full_checkpoint(latest_checkpoint.sequence_number)
+        .await
+        .is_err()
+    {
+        total_remaining_attempts -= 1;
+        if total_remaining_attempts == 0 {
+            return Ok(false);
+        }
+        tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+    }
```
Comment (halfprice) [inline crates/walrus-event/src/lib.rs:216]: Nit: there is a utils::retry() helper you can use. There is currently only one strategy, ExponentialBackoff, but I guess you can either set min and max backoff to be the same value, or create a FixedIntervalBackoff.

File: crates/walrus-event/src/lib.rs [hunk 2]
```diff
@@ -196,3 +196,30 @@ pub async fn get_bootstrap_committee_and_checkpoint(
     let verified_checkpoint = VerifiedCheckpoint::new_unchecked(checkpoint_data.checkpoint_summary);
     Ok((committee, verified_checkpoint))
 }
+
+async fn check_experimental_rest_endpoint_exists(client: Client) -> anyhow::Result<bool> {
+    // TODO: Once full nodes update their openapi.json spec to include the experimental endpoint, we
```
Comment (halfprice) [inline crates/walrus-event/src/lib.rs:201]: Suggestion: for TODO like this that there isn't an immediate action needed, I think typically we need to create a issue and put the issue number here to track it.

File: crates/walrus-event/src/lib.rs [hunk 3]
```diff
@@ -196,3 +196,30 @@ pub async fn get_bootstrap_committee_and_checkpoint(
...
+    }
+    Ok(true)
+}
+
+pub async fn ensure_experimental_rest_endpoint_exists(client: Client) -> anyhow::Result<()> {
+    if !check_experimental_rest_endpoint_exists(client.clone()).await? {
+        bail!("Full node does not support experimental endpoint");
+    }
+    Ok(())
+}
```
Comment (halfprice) [inline crates/walrus-event/src/lib.rs:225]: Nit: I feel we can just use `check_experimental_rest_endpoint_exists` in the call site. Adding another layer feels a bit unnecessary.

File: crates/walrus-event/src/lib.rs (hunk 1 above)
Comment (sadhansood) [reply crates/walrus-event/src/lib.rs:216]: That is a cool api, I just checked. But adding walrus-service to this crate causes cyclic dependency which requires some refactor to break. We should pull out the utils in another `walrus-utils` crate outside, but until then i'll keep this.

File: crates/walrus-event/src/lib.rs (hunk 2 above)
Comment (sadhansood) [reply crates/walrus-event/src/lib.rs:201]: Done

Comment (sadhansood) [reply crates/walrus-event/src/lib.rs:201]: I updated the todo in a different PR (sorry things got mixed up)

File: crates/walrus-event/src/lib.rs (hunk 1 above)
Comment (sadhansood) [reply crates/walrus-event/src/lib.rs:216]: updated the comment to use this api later once utils is outside walrus-service

File: crates/walrus-event/src/lib.rs (hunk 3 above)
Comment (sadhansood) [reply crates/walrus-event/src/lib.rs:225]: added an info log here as well which justifies this function now

File: crates/walrus-event/src/lib.rs [hunk 4]
```diff
@@ -196,3 +196,30 @@ pub async fn get_bootstrap_committee_and_checkpoint(
...
+            return Ok(false);
+        }
+        tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+    }
+    Ok(true)
+}
+
+pub async fn ensure_experimental_rest_endpoint_exists(client: Client) -> anyhow::Result<()> {
+    if !check_experimental_rest_endpoint_exists(client.clone()).await? {
+        bail!("Full node does not support experimental endpoint");
```
Comment (mlegner) [inline crates/walrus-event/src/lib.rs:222]: I'm a bit late, but one comment here: This error message may not be too helpful for the operator: Which experimental endpoint? What is it needed for?

Can we clarify this a bit?

Comment (sadhansood) [reply crates/walrus-event/src/lib.rs:222]: good call out, i can definitely see adding more details to this error message could be informative and avoid confusions.

Comment (mlegner) [reply crates/walrus-event/src/lib.rs:222]: Resolved in #1057.

**Pull request merged**